### ISSUE TITLE
Adapter/Handler does not execute enumerable data sequentially

### ISSUE THREAD

**Issue by goncalo-oliveira** [goncalo-oliveira/channels#17]
## Expected Behaviour

When an `IEnumerable<T>` is passed to an adapter/handler, it should be executed sequentially, from the first to the last. Otherwise data might be processed *out of order*.

## Current Behaviour

The adapter/handler is executing all elements of an enumerable on a separate task, coupling them with an `WhenAll` finalizing task. This may result in data being processed *out of order*.

## Possible Solution

I think this is a bug and as such, the solution should be to implement the expected behaviour. That is, executing items one by one, sequentially, from the first to the last.

## Workaround

To workaround this issue we can implement our adapter/handler with `T[]` or `IEnumerable<T>` instead of `T`. This way the entire array is always delivered and is up to the adapter to choose how to execute.

**Comment by goncalo-oliveira** [comment]
Don't think this should be considered a bug, but instead a design feature.

If the middleware is expecting `T` but receives `T[]` instead, the execution is spawned through multiple tasks. The *why* is to improve performance; the implementation is telling us that it wants to deal with each item individually and we provide a way of doing that in parallel.

If the sequence order is important, then the middleware should implement `T[]` instead and manually cycle through the elements.

The documentation should reflect that though.
